### PR TITLE
Implement detachChild for LayoutBuilder

### LINKED ISSUE #5840: State with global key gets disposed and recreated even if it remains in the tree with the same key.
With the following app, every time you tap you toggle the background decoration and Foo's FooState gets recreated.

```dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';

Future main() async {
  debugPrintBeginFrameBanner = true;
  debugPrintEndFrameBanner = true;
  runApp(new Bar());
}

class Bar extends StatefulWidget {
  @override
  BarState createState() => new BarState();
}

class BarState extends State<Bar> {
  bool _useBackground = true;

  @override
  Widget build(BuildContext context) => new GestureDetector(
        behavior: HitTestBehavior.opaque,
        onTap: () => setState(() {
              _useBackground = !_useBackground;
            }),
        child: new Container(
          decoration: !_useBackground
              ? null
              : new BoxDecoration(backgroundColor: Colors.black),
          child: new LayoutBuilder(
            builder: (_, __) => new Foo(key: _fooKey),
          ),
        ),
      );
}

final GlobalKey _fooKey = new GlobalKey();

class Foo extends StatefulWidget {
  Foo({Key key}) : super(key: key);
  @override
  FooState createState() => new FooState();
}

class FooState extends State<Foo> {
  FooState() {
    print('$hashCode FooState()');
  }

  @override
  void initState() {
    super.initState();
    print('$hashCode FooState.initState() ${config.key}');
  }

  @override
  void dispose() {
    super.dispose();
    print('$hashCode FooState.dispose() ${config.key}');
  }

  @override
  Widget build(_) => new Container();
}
```

Edit: Reduced a bit further.

### PR DESCRIPTION
**Pull request by Hixie**
Fixes https://github.com/flutter/flutter/issues/5840

### CODE REVIEW

**Comment by Hixie** [comment]
Also fixes https://github.com/flutter/flutter/issues/5834

File: packages/flutter/lib/src/widgets/virtual_viewport.dart [hunk 1]
```diff
@@ -100,6 +100,18 @@ abstract class VirtualViewportElement extends RenderObjectElement {
       visitor(child);
   }
 
+  @override
+  void detachChild(Element child) {
+    assert(() {
+      // TODO(ianh): implement detachChild for LazyBlock
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/virtual_viewport.dart]: for VirtualViewport...

**Comment by abarth** [comment]
LGTM

**Comment by Hixie** [comment]
fixed comment and lints, will check in when and if it goes green.

**Pull request merged**